### ISSUE TITLE
Deprecations will fire spuriously under strict ES module builds

### ISSUE THREAD

**Issue by ef4** [emberjs/data#9310]
Using ember-data 5.3.0 and given either:
 - the proposed [strict es modules RFC](https://github.com/emberjs/rfcs/pull/938) 
 - or an app built with the current Embroider main branch

ember-data emits deprecation warning for features that the app isn't actually using. 

```
DEPRECATION: You are relying on ember-data auto-magically installing the BooleanTransform. Use `export { BooleanTransform as default } from 'ember-data/serializer/transform';` in app/transforms/boolean.js instead [deprecation id: ember-data:deprecate-legacy-imports] This will be removed in ember-data 6.0.
DEPRECATION: You are relying on ember-data auto-magically installing the DateTransform. Use `export { DateTransform as default } from 'ember-data/serializer/transform';` in app/transforms/date.js instead [deprecation id: ember-data:deprecate-legacy-imports] This will be removed in ember-data 6.0.
DEPRECATION: You are relying on ember-data auto-magically installing the NumberTransform. Use `export { NumberTransform as default } from 'ember-data/serializer/transform';` in app/transforms/number.js instead [deprecation id: ember-data:deprecate-legacy-imports] This will be removed in ember-data 6.0.
DEPRECATION: You are relying on ember-data auto-magically installing the StringTransform. Use `export { StringTransform as default } from 'ember-data/serializer/transform';` in app/transforms/string.js instead [deprecation id: ember-data:deprecate-legacy-imports] This will be removed in ember-data 6.0.
```

Historically Ember has allowed "modules" to remain un-evaluated until they are synchronously required. That is not possible under true ES modules. So deprecations in module scope like here:

https://github.com/emberjs/data/blob/72b1e210091cdb59c61e911b50294b2a7fc941aa/packages/-ember-data/app/transforms/string.js#L5-L6

will fire before the app boots regardless of whether anybody is actually using that feature.

(One might also ask whether we can stop putting these modules in the build at all when they're not in use, and that's a good goal for the next-gen ember-data API, but these particular modules exist for backward compatibility and probably need to continue getting the backward-compatible behavior that pushes them into the build to make them synchronously available at runtime just in case they're used.)

**Comment by runspired** [comment]
> Historically Ember has allowed "modules" to remain un-evaluated until they are synchronously required. That is not possible under true ES modules.

This sounds like a bundler error then, not a spec error so I wouldn't phrase it as "under true ES modules". By spec, a module should not even be loaded until a module which requires it is evaluated, and so a spec compliant bundler would in fact not eagerly evaluate the module unless accessing a specific entry point ultimately led to accessing the file.

**Comment by ef4** [comment]
> By spec, a module should not even be loaded until a module which requires it is evaluated

Yes, but "requires" is an imprecise word here. What you mean is "imports it", because importing a module is the only legal way to depend on it. And ember-data doesn't ever import the transforms, which is the only reason we're having this discussion.

ember-data is going to do `getOwner(this).lookup('transform:boolean')` an expect to synchronously get back the contents of the ES module. That absolutely means that the module **must** have already been evaluated earlier, because there's no way to synchronously evaluate an ES module that wasn't already evaluated.

**Comment by runspired** [comment]
> ember-data is going to do getOwner(this).lookup('transform:boolean') an expect to synchronously get back the contents of the ES module

nit: a specific serializer we currently provide does do this though we consider it to be app code if an app makes use of it. Most of the time, it will be app code that does the resolver lookup. EmberData itself makes no use of Ember's resolver outside of adapterFor and serializerFor, both of which are also optional hooks and can be easily implemented to not use the resolver.

**Comment by runspired** [comment]
My general takeaway is the transform deprecation printing in this situation is desirable and not spurious, it just alerts the user to a requirement they may not have been paying attention to. If they truly don't want the code, then it might be time to consider bringing less of EmberData :D

**Comment by ef4** [comment]
In discussion today we agreed it would be helpful to change the timing of these deprecations to ease upgrades, so that they only fire for users who are actually using the transforms.

**Comment by wycats** [comment]
I hit this on a brand new Ember 6.x app using Vite and Warp Drive and I was surprised by it 😄
